src/arcaflow_plugin_sdk/atp.py

```diff
@@ -50,73 +64,143 @@ class HelloMessage:
 _HELLO_MESSAGE_SCHEMA = schema.build_object_schema(HelloMessage)
 
 
-def _handle_exit(_signo, _stack_frame):
-    print("Exiting normally")
-    sys.exit(0)
+def signal_handler(_sig, _frame):
+    pass  # Do nothing
 
 
-def run_plugin(
-    s: schema.SchemaType,
-    stdin: io.FileIO,
-    stdout: io.FileIO,
-    stderr: io.FileIO,
-) -> int:
-    """
-    This function wraps running a plugin.
-    """
-    if os.isatty(stdout.fileno()):
-        print("Cannot run plugin in ATP mode on an interactive terminal.")
-        return 1
+class ATPServer:
+    stdin: io.FileIO
+    stdout: io.FileIO
+    stderr: io.FileIO
+    step_object: typing.Any
 
-    signal.signal(signal.SIGTERM, _handle_exit)
-    try:
-        decoder = cbor2.decoder.CBORDecoder(stdin)
-        encoder = cbor2.encoder.CBOREncoder(stdout)
+    def __init__(
+            self,
+            stdin: io.FileIO,
+            stdout: io.FileIO,
+            stderr: io.FileIO,
+    ) -> None:
+        self.stdin = stdin
+        self.stdout = stdout
+        self.stderr = stderr
 
-        # Decode empty "start output" message.
-        decoder.decode()
+    def run_plugin(
+        self,
+        plugin_schema: schema.SchemaType,
+    ) -> int:
+        """
+        This function wraps running a plugin.
+        """
+        signal.signal(signal.SIGINT, signal_handler)  # Ignore sigint. Only care about arcaflow signals.
+        if os.isatty(self.stdout.fileno()):
+            print("Cannot run plugin in ATP mode on an interactive terminal.")
+            return 1
```

Comment:
The plugins communicate over stdout and stdin

Reply:
Understood, but does that mean that there should _never_ be a terminal connected there (not even for interactive testing)?  (Is it even possible that this could happen?)  That is, is there actually value to performing this check?

If so, then shouldn't you be checking `stdin` as well?